core/packages/contracts/contracts/NativeTokens.sol

```diff
@@ -0,0 +1,148 @@
+        /// @notice The destination address that will receive unlocked funds.
+        address recipient;
+        /// @notice The amount to unlock.
+        uint256 amount;
+    }
+
+    /// @notice Funds where locked.
+    /// @dev Emitted once the funds are locked and a message is successfully queued.
+    /// @param origin The address which initiated the lock.
+    /// @param recipient The substrate address that will receive the funds.
+    /// @param token The token locked.
+    /// @param amount The amount locked.
+    event Locked(address origin, bytes32 recipient, address token, uint256 amount);
+
+    /// @notice Funds where unlocked.
+    /// @dev Emitted once the funds are unlocked.
+    /// @param origin The substrate address which initiated the unlock.
+    /// @param recipient The ethereyn address that will receive the funds.
+    /// @param token The token unlocked.
+    /// @param amount The amount unlocked.
+    event Unlocked(bytes32 origin, address recipient, address token, uint256 amount);
+
+    /// @notice a token was created in Statemint.
+    /// @dev Emitted after enqueueing a a create token message to substrate.
+    /// @param token The address of the token created.
+    // TODO: Remove name, symbol and decimals.
+    event Created(address token, string name, string symbol, uint8 decimals);
+
+    /// @dev Zero amount.
+    error ZeroAmount();
+
+    /// @dev Zero addressed token.
+    error ZeroAddressToken();
+
+    /// @dev Zero addressed recipient.
+    error ZeroAddressRecipient();
+
+    /// @dev the origin provided is unauthorized.
+    error UnauthorizedOrigin();
+
+    /// @dev the message action is unsupported.
+    error UnsupportedMessageAction();
+
+    /// @dev the origin that
+    bytes32 public immutable allowedOrigin;
+
+    /// @dev The vault where ERC20 tokens are locked.
+    ERC20Vault public immutable vault;
+
+    /// @dev The channel used to enqueue messages for lock and create functions.
+    OutboundChannel public immutable outboundChannel;
+
+    /// @notice Initializes the NativeTokens contract with a vault and channels.
+    /// @param _vault The vault to use to `lock`/`unlock` tokens.
+    /// @param _outboundChannel The channel used to queue lock and create messages.
+    /// @param _allowedOrigin The origin allowed to call handle.
+    constructor(ERC20Vault _vault, OutboundChannel _outboundChannel, bytes32 _allowedOrigin) {
+        vault = _vault;
+        outboundChannel = _outboundChannel;
+        allowedOrigin = _allowedOrigin;
+    }
+
+    /// @notice Locks tokens to mint on substrate.
+    /// @dev Locks an amount of ERC20 Tokens in the vault and enqueues a mint message. Requires the allowance to be set on the ERC20 token where the spender is the vault.
+    /// @param token The token to lock.
+    /// @param recipient The recipient on the substrate side.
+    /// @param amount The amount to lock.
+    function lock(address token, bytes32 recipient, uint256 amount) public {
```

Comment:
Remember from the previous ERC20 app that users should only be allowed to lock `type(uint128).max` amount of tokens. 

Parachains on the Polkadot side don't support 256-bits of precision.

Reply:
Also we need to look back at your previous vault PR to see if we left the vault at 256-bits of precision. I think we did, but should confirm.

Reply:
It was mixed. I think I will leave the vault as uint256 and change `lock` and `unlock` code paths in NativeTokens to only use uint128 and validate this.

Reply:
Sounds good 👌